File: XCodeDemos/HelloMac/shader.cpp
//
//  shader.cpp
//  HelloMac
//
//  Created by Iann on 2019/6/12.
//  Copyright © 2019 iann. All rights reserved.
//
#define GL_SILENCE_DEPRECATION
#import <OpenGL/gl3.h>
#import <OpenGL/gl3ext.h>
#include "shader.hpp"

GLuint CreateShader(const char *vertShaderSource, const char *fragShaderSource)
{
    GLuint vertShaderId = glCreateShader(GL_VERTEX_SHADER);
    GLuint fragShaderId = glCreateShader(GL_FRAGMENT_SHADER);
    
    // Compile Vertex Shader
    glShaderSource(vertShaderId, 1, &vertShaderSource, 0);
    glCompileShader(vertShaderId);
    GLint status = GL_TRUE;
    GLint infoLen = 0;
    glGetShaderiv(vertShaderId, GL_COMPILE_STATUS, &status);
    glGetShaderiv(vertShaderId, GL_INFO_LOG_LENGTH, &infoLen);
    if (infoLen)
    {
        GLchar *buff = new GLchar[infoLen+1];
        buff[infoLen] = 0;
        glGetShaderInfoLog(vertShaderId, infoLen, 0, buff);
        printf("%s", buff);
        delete[] buff;
    }
    
    // Compile Fragment Shader
    glShaderSource(fragShaderId, 1, &fragShaderSource, 0);
    glCompileShader(fragShaderId);
    status = GL_TRUE;
    infoLen = 0;
    glGetShaderiv(fragShaderId, GL_COMPILE_STATUS, &status);
    glGetShaderiv(fragShaderId, GL_INFO_LOG_LENGTH, &infoLen);
    if (infoLen)
    {
        GLchar *buff = new GLchar[infoLen+1];
        buff[infoLen] = 0;
        glGetShaderInfoLog(fragShaderId, infoLen, 0, buff);
        printf("%s", buff);
        delete[] buff;
    }
    
    // Link the program
    GLuint programId = glCreateProgram();
    glAttachShader(programId, vertShaderId);
    glAttachShader(programId, fragShaderId);
    glLinkProgram(programId);
    status = GL_TRUE;
    infoLen = 0;
    glGetProgramiv(programId, GL_COMPILE_STATUS, &status);
    glGetProgramiv(programId, GL_INFO_LOG_LENGTH, &infoLen);
    if (infoLen)
    {
        GLchar  *buff = new GLchar[infoLen + 1];
        buff[infoLen] = 0;
        glGetProgramInfoLog(programId, infoLen, 0, buff);
        printf(buff);
        delete buff;
    }
    
    glDetachShader(programId, vertShaderId);
    glDetachShader(programId, fragShaderId);
    glDeleteShader(vertShaderId);
    glDeleteShader(fragShaderId);
    return programId;
}

File: AndroidDemos/Robot/app/src/main/java/com/example/robot/Shader.java
package com.example.robot;

import android.opengl.GLES20;
import android.util.Log;

public class Shader
{
    public static int LoadShaderFromSourceCode(String vertexShaderCode, String fragmentShaderCode)
    {
        int vertexShader = GLES20.glCreateShader(GLES20.GL_VERTEX_SHADER);
        int fragmentShader = GLES20.glCreateShader(GLES20.GL_FRAGMENT_SHADER);

        final int[] compileStatus = new int[1];

        // Compiling vertex shader.
        GLES20.glShaderSource(vertexShader,vertexShaderCode);
        GLES20.glCompileShader(vertexShader);
        GLES20.glGetShaderiv(vertexShader, GLES20.GL_COMPILE_STATUS, compileStatus, 0);
        if (compileStatus[0] == 0)
        {
            Log.v("Robot", "Results of compiling source:" + "\n" + vertexShaderCode + "\n:"
                    + GLES20.glGetShaderInfoLog(vertexShader));
        }



        // Compiling fragment shader.
        GLES20.glShaderSource(fragmentShader,fragmentShaderCode);
        GLES20.glCompileShader(fragmentShader);
        GLES20.glGetShaderiv(fragmentShader, GLES20.GL_COMPILE_STATUS, compileStatus, 0);
        if (compileStatus[0] == 0)
        {
            Log.v("Robot", "Results of compiling source:" + "\n" + fragmentShaderCode + "\n:"
                    + GLES20.glGetShaderInfoLog(fragmentShader));
        }

        final int program = GLES20.glCreateProgram();
        GLES20.glAttachShader(program, vertexShader);
        GLES20.glAttachShader(program, fragmentShader);
        GLES20.glLinkProgram(program);


        final int[] linkStatus = new int[1];
        GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, 0);


        return program;
    }
}
